packages/test/test-utils/src/timeoutUtils.ts

```diff
@@ -4,15 +4,120 @@
  */
 
 import { Container } from "@fluidframework/container-loader";
+import { assert } from "@fluidframework/common-utils";
 
+// @deprecated this value is no longer used
 export const defaultTimeoutDurationMs = 250;
 
+// TestTimeout class manage tracking of test timeout. It create a timer when timeout is in effect,
+// and provide a promise that will be reject before the test timeout happen with a `timeBuffer` of 15 ms.
+// Once rejected, a new TestTimeout object will be create for the timeout.
+
+type PromiseRejectFunc = (reason?: any) => void;
+const timeBuffer = 15; // leave 15 ms leeway for finish processing
+
+class TestTimeout {
+    private timeout: number = 0;
+    private timer: NodeJS.Timeout | undefined;
+    private reject: PromiseRejectFunc = () => { };
+    private readonly promise: Promise<void>;
+    private rejected = false;
+
+    private static instance: TestTimeout;
+    public static initialize() {
+        TestTimeout.instance = new TestTimeout();
+    }
+    public static reset(runnable: Mocha.Runnable) {
+        TestTimeout.clear();
+        TestTimeout.instance.resetTimer(runnable);
+    }
+
+    public static clear() {
+        if (TestTimeout.instance.rejected) {
+            TestTimeout.instance = new TestTimeout();
+        } else {
+            TestTimeout.instance.clearTimer();
+        }
+    }
+
+    public static getInstance() {
+        return TestTimeout.instance;
+    }
+
+    public async getPromise() {
+        return this.promise;
+    }
+
+    public getTimeout() {
+        return this.timeout;
+    }
+
+    private constructor() {
+        this.promise
+            = new Promise((resolve, reject: PromiseRejectFunc) => { this.reject = reject; });
+        // Ignore rejection for timeout promise if no one is waiting for it.
+        this.promise.catch(() => { });
+    }
+
+    private resetTimer(runnable: Mocha.Runnable) {
+        assert(!this.timer, "clearTimer should have been called before reset");
+        assert(!this.rejected, "can't reset a rejected TestTimeout");
+
+        // Check the test timeout setting
+        const timeout = runnable.timeout();
+        if (!(Number.isFinite(timeout) && timeout > 0)) { return; }
+
+        // subtract a buffer
+        this.timeout = Math.max(timeout - timeBuffer, 1);
+
+        // Set up timer to reject near the test timeout.
+        this.timer = setTimeout(() => {
```

Comment:
this feels like a lot of complexity to just create a timer. i think i'd prefer to just make timeoutPromise smarter, and pull the timeout off the test in the default case

Reply:
i'm hoping we could just do something like `timeout = runnable.duration - runnable.timeout() - timeBuffer`

Reply:
if not. the below code in `if (globalThis.getMochaModule !== undefined)` could just set some var to be the start time in  runnablePrototype.resetTimeout  which would take the role of runnable.duration above

Reply:
`resetTimeout` get calls after the async test function returns the promise, it is too late to influence the time `timeoutPromise` uses

Reply:
Also, capture the end time was my original solution.  But that doesn't work if the call to set the timeout is within the test itself.   So we will have reset/clear/reset/clear.   We need to have a way to cancel and change the time out for `timeoutPromise` that is already in flight.

Reply:
No. mocha parallelism is achieved via different process.